### PR TITLE
Shopping: Fix Shopify product queries, add tests

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

I noticed in my testing that Shopify queries were broken since #11340, because the code tried to save `Product` instances in transients instead of something serializable.

## Summary

In my attempt to fix the Shopify product queries I refactored the class and added a bunch of tests to make things more robust and also attempt to do JSON schema validation on the PHPUnit side.

## Relevant Technical Choices

* New `assertMatchesProductSchema` matcher for JSON schema validation
* Wanted a way to test that the `Product`s returned by `Shopify_Query` adhere to the JSON schema, figured that `JsonSerializable` would work fine for that.

Patches the product schema for https://github.com/ampproject/amphtml/pull/38214

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Related:

* #11154 
* #11210

### CODE REVIEW

File: packages/wp-dashboard/src/components/editorSettings/utils/validateShopifyHost.js [hunk 1]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-export const hostPattern = /[\d\w-]+\.myshopify\.com/i;
+export const hostPattern = /[\w-]+\.myshopify\.com/i;
```
Comment (swissspidy) [inline packages/wp-dashboard/src/components/editorSettings/utils/validateShopifyHost.js:17]: This was redundant.

The `\w` metacharacter already matches digits

File: packages/e2e-tests/src/specs/editor/shopping/schema.json [hunk 2]
```diff
@@ -216,6 +216,7 @@
         "type": "object",
         "properties": {
           "url": {
+            "type": "string",
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/schema.json:219]: Patch from https://github.com/ampproject/amphtml/pull/38214

File: tests/phpunit/integration/tests/User/Preferences.php [hunk 3]
```diff
@@ -193,7 +193,8 @@ public function test_permission_check_for_authors(): void {
 			$response = $response->as_error();
 		}
 
-		$this->assertWPError( $response, 'rest_cannot_edit' );
+		$this->assertWPError( $response );
```
Comment (swissspidy) [inline tests/phpunit/integration/tests/User/Preferences.php:196]: This was incorrect usage of `assertWPError`

File: includes/Shopping/Product.php [hunk 4]
```diff
@@ -83,11 +85,7 @@ class Product {
 	 *
 	 * @var array
 	 */
-	protected $aggregate_rating = [
-		'rating_value' => 0.0,
-		'review_count' => 0,
-		'review_url'   => '',
-	];
+	protected $aggregate_rating;
```
Comment (swissspidy) [inline includes/Shopping/Product.php:88]: An "empty" rating array like this is invalid. If provided, there _must_ be a URL as well.

So if there's no rating information, this needs to be omitted.

File: includes/Shopping/Shopify_Query.php [hunk 5]
```diff
@@ -59,76 +60,105 @@ public function __construct( Settings $settings ) {
...
-		 * @param int $time Time to live (in seconds). Default is 1 day.
-		 */
-		$cache_ttl = apply_filters( 'web_stories_shopify_data_cache_ttl', DAY_IN_SECONDS ); // TODO: lower default TTL.
-		$cache_key = ! empty( $search_term ) ? 'web_stories_shopify_data_' . md5( $search_term ) : 'web_stories_shopify_data_default';
+		if ( ! preg_match( '/^[\w-]+\.myshopify\.com$/i', $host ) ) {
+			return new WP_Error( 'rest_invalid_hostname', __( 'Invalid Shopify hostname.', 'web-stories' ), [ 'status' => 400 ] );
+		}
 
-		$data = get_transient( $cache_key );
+		$url = esc_urL_raw(
```
Comment (spacedmonkey) [inline includes/Shopping/Shopify_Query.php]: Is this a mistake?

File: includes/Shopping/Product.php [hunk 6]
```diff
@@ -26,10 +26,12 @@
 
 namespace Google\Web_Stories\Shopping;
 
+use JsonSerializable;
+
 /**
  * Class Product
  */
-class Product {
+class Product implements JsonSerializable {
```
Comment (spacedmonkey) [inline includes/Shopping/Product.php:34]: What is this used for and why is it needed?

**Comment by spacedmonkey** [comment]
This looks great! Couple of questions /  nicpics I will approve.

File: includes/Shopping/Product.php (hunk 6 above)
Comment (swissspidy) [reply includes/Shopping/Product.php:34]: See PR description. This is what I mentioned in our sync yesterday.

I wanted a way to verify that the Shopify provider returns valid products, so I came up with this idea to serialize each `Product` to JSON and then validate that against the schema.

File: includes/Shopping/Shopify_Query.php (hunk 5 above)
Comment (swissspidy) [inline includes/Shopping/Shopify_Query.php]:
```suggestion
		$url = esc_url_raw(
```

**Review by timarney — APPROVED** [review]
Tested locally 👍

**Comment by timarney** [comment]
Just noticed you might want to use `priceRangeV2 ` in the query

https://github.com/GoogleForCreators/web-stories-wp/pull/11519/files#diff-4b687062d3e63e8611fe3e309f0a009bf802275ae5c9a6bb4af2bd6ced6df50cR171

Screen Shot 2022-05-20 at 8 57 47 PM

**Pull request merged**